plugins/org.eclipse.elk.alg.vertiflex/src/org/eclipse/elk/alg/vertiflex/p2relative/RelativeXPlacer.java

```diff
@@ -0,0 +1,651 @@
+                    double rightOutlineX, posX;
+                    for (int i = 0; i < maxDepthStartPos; i++) {
+                        for (int j = i + 1; j < maxDepthStartPos + 1; j++) {
+                            rightOutline = children.get(i).getProperty(InternalProperties.RIGHT_OUTLINE);
+                            rightOutlineX = children.get(i).getX() + rightOutline.getRelativeX();
+                            posX = children.get(j).getX() + children.get(j).getWidth() / 2.0;
+                            while (rightOutline != null && rightOutline.getAbsoluteY() < maxDepth) {
+    
+                                newMoveRoot = posX - graph.getWidth() / 2.0 + (posX - rightOutlineX) * ((graph.getY() 
+                                        + graph.getHeight()) - maxDepth) / (maxDepth - rightOutline.getAbsoluteY());
+                                betterMoveRoot = Math.max(betterMoveRoot, newMoveRoot);
+    
+                                rightOutline = rightOutline.getNext();
+                                if (rightOutline != null) {
+                                    rightOutlineX += rightOutline.getRelativeX();
+                                }
+                            }
+                        }
+                    }
+                    moveRoot = betterMoveRoot;
+                }
+    
+                if (betterMoveRoot > moveRoot) {
+                    OutlineNode leftOutline;
+                    double leftOutlineX, posX;
+                    for (int i = pos; i < children.size(); i++) {
+                        for (int j = pos - 1; j < i; j++) {
+                            leftOutline = children.get(i).getProperty(InternalProperties.LEFT_OUTLINE);
+                            leftOutlineX = children.get(i).getX() + leftOutline.getRelativeX();
+                            posX = children.get(j).getX() + children.get(j).getWidth() / 2.0;
+                            while (leftOutline != null && leftOutline.getAbsoluteY() < maxDepth) {
+    
+                                newMoveRoot = posX - graph.getWidth() / 2.0 + (posX - leftOutlineX) * ((graph.getY() 
+                                        + graph.getHeight()) - maxDepth) / (maxDepth - leftOutline.getAbsoluteY());
+                                betterMoveRoot = Math.min(betterMoveRoot, newMoveRoot);
+    
+                                leftOutline = leftOutline.getNext();
+                                if (leftOutline != null) {
+                                    leftOutlineX += leftOutline.getRelativeX();
+                                }
+                            }
+                        }
+                    }
+                    moveRoot = betterMoveRoot;
+                }
+            }
+            
+            for (ElkNode child: children) {
+                child.setX(child.getX() - moveRoot);
+            }
+            
+            
+            double newX;
+            // left outline update
```

Comment:
Maybe reduce map access here by saving some of the properties instead of requesting them multiple times.
See also below